### DISCUSSION TITLE
How to Pass Commandline Arguments and Properties

### DISCUSSION THREAD

**Discussion by schrieveslaach** [mfussenegger/nvim-dap#536] [Q&A] [answered]
Is it possible to pass command line arguments to [Java configurations](https://github.com/mfussenegger/nvim-dap/wiki/Java)? And is it possible to pass JVM properties?

**Comment by mfussenegger** [comment, marked as answer]
You can add `args` and `vmArgs`. See https://github.com/microsoft/java-debug/blob/625fe84cef4825a4b6b476d04383f5168e81708d/com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/protocol/Requests.java#L104-L119

Not sure if there is also proper documentation for it somewhere.

Here is an example of it being used: https://github.com/crate/crate/blob/e1dea641515e57f61e77dcd34d8b2de042878bbf/.vscode/launch.json#L4-L13

And I recommend to use it with nvim-jdtls, otherwise you need to also define the full classpath and so on.

**Reply by mfussenegger** [reply]
Created a PR to document them: https://github.com/microsoft/java-debug/pull/412

**Comment by schrieveslaach** [comment]
Thanks. I did not find it in the documentation (not in nvim-jdtls' or in nvim-dap) so i wasn't aware of these options.

A follow-up question: is it possible to configure them for a one shot start? I could imaging that `continue()` has a list of args for one shot debugging, e.g. `require('dap').continue({ args: {"a", "list", "of",  "args"}})`

**Reply by mfussenegger** [reply]
You can either use a `function` as value in the configuration (if done in lua) and use inputlist/input to prompt for arguments ad-hoc or otherwise you have to call `require('dap').run(config)` with an ad-hoc configuration that you created.

**Comment by abenz1267** [comment]
Just gonna hijack this because it's somewhat related: i'm trying to get nvim-dap to work with jmockit, apparently as of JMockit 1.42 it requires a `-javaagent` jvm init parameter (see also here https://jmockit.github.io/tutorial/Introduction.html#runningTests).

I've tried adding it like this:

```lua
    dap.configurations.java = {
        request = "launch",
        type = "java",
        vmArgs = "-javaagent:/home/andrej/.m2/repository/org/jmockit/jmockit/1.49/jmockit-1.49.jar"
    }
```

Is this correct? I can't get it to work properly. Tests run fine via Maven. A dependency annotated with `@Tested` stays null when running in nvim and i suspect it's because of the `-javaagent` setting.

Regards

**Reply by schrieveslaach** [reply]
I think you need to hook into `require('jdtls').test_class()` because that will create a Lua table that will be passed to `require('dap').run(config_generated_by_jdtls_test_class)`. I'm not sure how you can do this but I hooked into nvim-jdtls to modify the run configuration:

```lua
local M = {}

M.run_with_extra_jvm_args = function()
   local dap = require('dap')
   local configurations = dap.configurations or {}

   if configurations.java then
      vim.ui.select(configurations.java, {
         prompt = 'Select Java Main',
         format_item = function(item)
            return item.name
         end,
      }, function(choice)
         local config = vim.deepcopy(choice)
         config.vmArgs = 'My custom choice'
         dap.run(config)
      end)
   else
      print('No Java configurations available. Please, run JdtRefreshDebugConfigs')
   end
end

return M
```

**Reply by mfussenegger** [reply]
You can pass `config_overrides` if you call `jdtls.test_class` or if you call `setup_dap_main_class_configs` / `fetch_main_configs`.

E.g.:

```lua
require('jdtls.dap').setup_dap_main_class_configs({ config_overrides = { vmArgs = 'my-custom-args' }})
```

Should work.
https://github.com/mfussenegger/nvim-jdtls/pull/277 also added some API reference documentation.
